### ISSUE TITLE
snrm2 overflow for large input floats

### ISSUE THREAD

**Issue by rth** [flame/blis#332]
DSNRM2 in BLIS overflows when applied to a an array `[1e20]` while it doesn't overflow in the Netlib BLAS (or OpenBLAS).

I run into this while trying to pass the scipy test suite with BLIS in https://github.com/scipy/scipy/issues/10744, as mentionned in that issue,

> Netlib BLAS does indeed [apply a scale in SNRM2](http://www.netlib.org/lapack/explore-html/df/d28/group__single__blas__level1_ga35c2ec0e9bfdaa545320c2a134fcc471.html#ga35c2ec0e9bfdaa545320c2a134fcc471) to avoir the overflow, but I don't know it it's actually part of the BLAS spec.

There are also a number of other scipy test failures due to overflow I think in other functions (see the linked issue).

**Comment by rth** [comment]
(Sorry submitted too early by accident, see edited description above. This issue is related to overflows when trying to pass the scipy test suite with BLIS).

**Comment by devinamatthews** [comment]
Possible solutions:

1) Cast f32 to f64 before multiplication and accumulate in f64 (looks like OpenBLAS does this maybe?).
2) Scale:
    - By using `isamax` to find the largest element.
    - Using `isamax` over chunks that fit in cache and normalizing each chunk individually.
    - Changing the scale factor as you go along (like netlib), but of course this isn't (easily) vectorizable.

    In each case the scale factor should be a power of two to avoid explicit division.

**Comment by rth** [comment]
Indeed. I imagine one would pay for it with performance in any case...

> Scale
> By using isamax to find the largest element.

Maybe the scale could only to be applied if `max_value**2*n_elements>dtype.max` to avoid needless calculations? Or some other way of detecting the overflow.

**Comment by devinamatthews** [comment]
I think the extra calculation is probably not an issue (except division but that can be avoided). The biggest cost would be either the increased memory accesses (going through the array twice) or the loss of vectorization. My preference is method 1) for `snrm2` and 2b) for `dnrm2`. @fgvanzee ?

**Comment by fgvanzee** [comment]
This is kind of beyond my preferred level of expertise, and also hard to follow without more details / code samples. I nominate @devinamatthews to throw together a pull request.

**Comment by devinamatthews** [comment]
@fgvanzee can you give me a list of the files I need to edit? I'm not sure I could find them all.

**Comment by fgvanzee** [comment]
`?nrm2()`, defined in `frame/compat/bla_nrm2.c`, is currently based on `bli_?normfv()`, which is defined in `frame/util/bli_util_tapi.c`, but is based on a helper function that is defined starting on line 313 in `frame/util/bli_util_unb_var1.c`. (Be sure to mind the `#ifdef FE_OVERFLOW` cpp branch.) That helper function currently uses another helper function that implements `sumsqv`, which is defined starting at line 1118 in `frame/util/bli_util_unb_var1.c`.

**Comment by devinamatthews** [comment]
So the functions there can't have optimized implementations? Not that it really matters for nrm2 and similar but I just want to make sure this is the only place to change.

**Comment by rvdg** [comment]
We have an enrichment on this in the course we are creating: 

http://www.cs.utexas.edu/users/flame/laff/alaff-beta/chapter01-computing-2-norm.html <http://www.cs.utexas.edu/users/flame/laff/alaff-beta/chapter01-computing-2-norm.html>

**Comment by fgvanzee** [comment]
> So the functions there can't have optimized implementations?

As those operations are currently classified, correct. Let me know if it becomes and problem and I can think about reclassifying those functions as level-1v, at which time they will "first-class citizens" of the `cntx_t`.

**Comment by devinamatthews** [comment]
@rth by any chance were you testing on a Mac? I looked at the code and in fact it is already applying a scale factor, except that it tries to optimize by computing dot(x,x) first and checking for overflow exceptions. On MacOS, though, floating point exceptions are never generated.

**Comment by rth** [comment]
>  by any chance were you testing on a Mac?

No, I was testing on Linux (Gentoo), using blis 0.6.0 though -- maybe this was recently changed?

**Comment by devinamatthews** [comment]
It looks like the behavior has been the same since Nov. 2016. I pushed in a new commit to #334 that unconditionally disables the use of dot. See if that works for you.

**Comment by rth** [comment]
I haven't build BLIS from sources yet (and won't be able to look into it before next week).

Are you saying that you are not able to reproduce the failure on master with a unit test applying snrm2 on an array [1e20] of dtype float (32 bit)?

**Comment by devinamatthews** [comment]
@field can you try and reproduce this and see if the fix in #334 works? @rth is this an array of length 1e20? Are the values drawn from some distribution?

**Comment by fgvanzee** [comment]
@devinamatthews I think @rth means that the array has one element containing the value 1e20.

Yeah, I can try to reproduce.

**Comment by rth** [comment]
>  is this an array of length 1e20? Are the values drawn from some distribution?

Sorry for the confusion. It's a 1D array with a single element equal to 1e20, so that when you square it, float32 would overflow.

**Comment by isuruf** [comment]
I can reproduce this issue on linux and can confirm that #334 works

**Comment by devinamatthews** [comment]
@fgvanzee saved by the bell. I think merging #334 is the ticket then, unless you want to clean it up/add a comment or something.

**Comment by fgvanzee** [comment]
I also can confirm that `snrm2_()` on 1e20 doesn't work. (I actually get `-nan`.)

> I think merging #334 is the ticket then, unless you want to clean it up/add a comment or something.

I'll take a look at it. Thanks for your quick work, everyone.

**Comment by rth** [comment]
For future reference, here is a python snipped reproducing the original issue,
```py
import ctypes
from ctypes import byref, c_int
import numpy as np

_blas = ctypes.cdll.LoadLibrary("lib/haswell/libblis.so")


def snrm2(x):

    _blas.snrm2_.restype = ctypes.c_float
    return _blas.snrm2_(
        byref(c_int(len(x)))),
        x.ctypes.data_as(ctypes.c_void_p),
        byref(c_int(1)))


x = np.array([2e-28], dtype=np.float32)
# should not return 0.0
out = snrm2(x)
print(out)

x = np.array([2e20], dtype=np.float32)
# should not return inf
out = snrm2(x)
print(out)
```

and I can confirm that https://github.com/flame/blis/pull/334 fixes it.
